src/Sparrow/Json/BlittableJsonReaderVector.cs

```diff
@@ -0,0 +1,338 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Sparrow.Json.Parsing;
+using static Sparrow.PortableExceptions;
+
+namespace Sparrow.Json
+{
+    public enum BlittableVectorType : byte
+    {
+        SByte = 0b0000_0001,  // 1 byte
+        Int16 = 0b0000_0010,  // 2 bytes
+        Int32 = 0b0000_0100,  // 4 bytes
+        Int64 = 0b0000_1000,  // 8 bytes
+
+        Byte = 0b1000_0001,  // 1 byte
+        UInt16 = 0b1000_0010,  // 2 bytes
+        UInt32 = 0b1000_0100,  // 4 bytes
+        UInt64 = 0b1000_1000,  // 8 bytes
+        
+        Half =   0b1100_0010, // 2 bytes
+        Float =  0b1100_0100, // 4 bytes
+        Double = 0b1100_1000, // 8 bytes
+    }
+
+
+    [StructLayout(LayoutKind.Explicit, Size = 8)]
+    internal struct BlittableVectorHeader
+    {
+        private const byte TypeMask = 0b0011_1111;
+        private const byte FloatFlag = 0b0100_0000;
+        private const byte UnsignedFlag = 0b1000_0000;
+
+        [FieldOffset(0)]
+        private BlittableVectorType _type;
+
+        [FieldOffset(1)]
+        public byte AlignmentOffset;
+
+        [FieldOffset(2)]
+        public int Count;
+
+        public BlittableVectorHeader(BlittableVectorType type, int count)
+        {
+            _type = type;
+            Count = count;
+        }
+
+        public BlittableVectorType Type => _type;
+
+        public bool IsFloatingPoint => (BlittableVectorType)((byte)_type & FloatFlag) != 0;
+
+        public bool IsUnsigned => (BlittableVectorType)((byte)_type & UnsignedFlag) != 0;
+
+        public int ElementSize => ((byte)Type) & TypeMask;
+        
+        
+    }
+    
+    public sealed unsafe class BlittableJsonReaderVector : BlittableJsonReaderBase, IEnumerable
+    {
+        private readonly BlittableVectorHeader* _header;
+        private readonly byte* _dataStart;
+
+        public int Length => _header->Count;
+
+        public BlittableVectorType Type => _header->Type;
+
+        public BlittableJsonReaderVector(byte* mem, int bufferSize, JsonOperationContext context)
+            : base(context)
+        {
+            //otherwise SetupPropertiesAccess will throw because of the memory garbage
+            //(or won't throw, but this is actually worse!)
+            ThrowIf<ArgumentException>(bufferSize == 0, $"{nameof(BlittableJsonReaderObject)} does not support objects with zero size");
+
+            _mem = mem;
+            _header = (BlittableVectorHeader*)mem; // +1 to skip the token byte
+            _dataStart = mem + sizeof(BlittableVectorHeader) + _header->AlignmentOffset;
+        }
+
+        public object this[int i]
```

Comment:
I implemented because it does exist, but I kinda disabled that usage with the `allow compression` defaulting to `false`. PGO would get rid of the boxing for selected codepaths.